lib/remap-lang/src/type_def.rs

```diff
@@ -167,7 +167,13 @@ impl TypeDef {
     }
 
     pub fn with_constraint(mut self, kind: impl Into<value::Kind>) -> Self {
-        self.kind = kind.into();
+        let kind = kind.into();
+
+        if kind.is_scalar() {
+            self.inner_type_def = None;
+        }
+
+        self.kind = kind;
```

Comment:
```suggestion
        self.kind = kind.into();

        if self.kind.is_scalar() {
            self.inner_type_def = None;
        }
```

Slightly more concise IMO.